src/providers/mdal/qgsmdalprovider.cpp

```diff
@@ -48,11 +48,23 @@ QgsCoordinateReferenceSystem QgsMdalProvider::crs() const
   return mCrs;
 }
 
+QStringList QgsMdalProvider::subLayers() const
+{
+  return mSubLayersUris;
+}
+
 QgsMdalProvider::QgsMdalProvider( const QString &uri, const ProviderOptions &options )
   : QgsMeshDataProvider( uri, options )
 {
   temporalCapabilities()->setTemporalUnit( QgsUnitTypes::TemporalHours );
-  loadData();
+  QByteArray curi = dataSourceUri().toUtf8();
+  bool hasMeshName = uri.split( ':' ).count() == 3;
```

Comment:
again, this is not that easy. win drivers have C:\ for example...

why we need this bool hasMeshName at all? you can pass the whole string to MDAL_MeshNames and it will parse it (even if it has driver name or mesh name) ...?

Reply:
Agree for ":" and windows files...

MDAL_MeshNames( curi ) return always all the mesh uris of the whole file, so to distinguish the case when the input URI are the proper MDAL URI, "driver:"file name":meshName", and the case of the input URI has another form, that's means we have to use MDAL_MeshNames( curi ) to parse the file.
But sure, I agree, this logic is not easy to read in this code.

Maybe it is better to use 
```
if (uri.split( ':' ).count() == 3)
{
loadData();
return;
}
QByteArray curi = dataSourceUri().toUtf8();
QStringList meshNames = QString( MDAL_MeshNames( curi ) ).split( QStringLiteral( ";;" ) );

 if ( meshNames.count() == 1 )
   loadData();
 else
   mSubLayersUris = meshNames;
```
But still has the windows file issue...
Any other idea to know if the uri is the proper MDAL uri with mesh name ?

Reply:
Maybe with a flag in the `QgsMeshLayer::LayerOptions` to tell the provider that this is a uri of a single mesh ?

Reply:
Ok, resolved with testing if containing "\":".